Commit b09e87e0 causes link error "undefined reference to `pthread_create'" on Linux

On Linux, SeqAn seems to need libpthread for multithreaded BAM file I/O (at least).  But since commit b09e87e0 (I think; I haven't tried git bisect to verify), using any SeqAn code that needs libpthread leads to the linker error "undefined reference to `pthread_create'".

TTBOMK, the reason is: Prior to commit b09e87e0, libpthread was included by appending `pthread` to `SEQAN_LIBRARIES` in CMakeLists.txt, which has the effect of *appending* `-lpthread` to the command line sent to the linker.  Since that commit, instead of appending to `SEQAN_LIBRARIES`, some linker options including `-lpthread` are appended to `CMAKE_EXE_LINKER_FLAGS`, which has the effect of *prepending* `-lpthread` to the command line sent to the linker -- and since the linker (on Linux anyway) only extracts library object code for symbols referenced by object files *that appeared earlier on the linker command line*, this means that `pthread_create` is never found, causing the link failure.  (FWIW this linker behaviour seems crazy to me.  But perhaps there's some justification if you dig deep enough...)

What puzzles me is why this build error apparently hasn't caused problems with SeqAn's own CI tests.  I'm running a very standard Ubuntu Linux 4.8.0-59-generic with g++ 6.3.0, so I don't think it's some peculiarity of my system.  I'm on the `develop` branch.

Note that prior to this commit, the "new" behaviour of using `CMAKE_EXE_LINKER_FLAGS` instead of `SEQAN_LIBRARIES` to include libpthread could already occur under other conditions (when building statically, it would seem), and it was possibly broken for that also.

**Workaround:** Manually specify `SEQAN_LIBRARIES` when running CMake.  For me this amounted to adding `-DSEQAN_LIBRARIES="pthread;rt;/usr/lib/x86_64-linux-gnu/libz.so"` to the CMake command line.

Hi, thanks for getting back to me.

I'm referring to building a simple application of my own, using a few-days-old `develop` branch of SeqAn with the 2-line fix in https://github.com/seqan/seqan/pull/2266 on top of it.  I'm using CMake on a reasonably recent Ubuntu inside a VirtualBox on Windows:

```
wtwhite@wtwhite-VirtualBox:~$ uname -a
Linux wtwhite-VirtualBox 4.8.0-59-generic #64-Ubuntu SMP Thu Jun 29 19:38:34 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
wtwhite@wtwhite-VirtualBox:~$ g++ --version
g++ (Ubuntu 6.3.0-12ubuntu2) 6.3.0 20170406
Copyright (C) 2016 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

wtwhite@wtwhite-VirtualBox:~$ cmake --version
cmake version 3.7.2

CMake suite maintained and supported by Kitware (kitware.com/cmake).
```

It sounds like you guys are having no problems with builds so it could certainly be something I'm doing, but I can't think what...  For comparison, there were no link problems when I was using the `libseqan2-dev` package version 2.3.1+dfsg-3ubuntu1 installed by `apt-get`.

I don't recall where I read about the way the linker looks up symbols unfortunately, but I definitely remember banging my head against this problem years ago.

I've attached a MWE, including CMakeLists.txt, that reproduces the link error for me if I have commit f3bef57 checked out under `$SEQAN` and run `mkdir build && cd build && cmake -DCMAKE_PREFIX_PATH=$SEQAN/util/cmake -DSEQAN_INCLUDE_PATH=$SEQAN/include .. && make VERBOSE=1`.  The output I get from this is:

```
wtwhite@wtwhite-VirtualBox:~/win/repro_seqan_pthread_link_error$ mkdir build && cd build && cmake -DCMAKE_PREFIX_PATH=$SEQAN/util/cmake -DSEQAN_INCLUDE_PATH=$SEQAN/include .. && make VERBOSE=1
-- The CXX compiler identification is GNU 6.3.0
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
CMake build type: Release
-- Found ZLIB: /usr/lib/x86_64-linux-gnu/libz.so (found version "1.2.8") 
-- Performing Test CXX14_BUILTIN
-- Performing Test CXX14_BUILTIN - Success
-- Looking for C++ include execinfo.h
-- Looking for C++ include execinfo.h - found
-- Found Seqan: /home/wtwhite/win/seqan/include (found version "2.4.0")
SEQAN_LIBRARIES=<rt;/usr/lib/x86_64-linux-gnu/libz.so>
-- Configuring done
-- Generating done
-- Build files have been written to: /home/wtwhite/win/repro_seqan_pthread_link_error/build
make: Warning: File 'Makefile' has modification time 1,4 s in the future
/usr/bin/cmake -H/home/wtwhite/win/repro_seqan_pthread_link_error -B/home/wtwhite/win/repro_seqan_pthread_link_error/build --check-build-system CMakeFiles/Makefile.cmake 0
/usr/bin/cmake -E cmake_progress_start /home/wtwhite/win/repro_seqan_pthread_link_error/build/CMakeFiles /home/wtwhite/win/repro_seqan_pthread_link_error/build/CMakeFiles/progress.marks
make -f CMakeFiles/Makefile2 all
make[1]: Entering directory '/home/wtwhite/win/repro_seqan_pthread_link_error/build'
make[1]: Warning: File 'CMakeFiles/Makefile2' has modification time 1,2 s in the future
make -f CMakeFiles/x.dir/build.make CMakeFiles/x.dir/depend
make[2]: Entering directory '/home/wtwhite/win/repro_seqan_pthread_link_error/build'
make[2]: Warning: File 'CMakeFiles/x.dir/flags.make' has modification time 1,2 s in the future
cd /home/wtwhite/win/repro_seqan_pthread_link_error/build && /usr/bin/cmake -E cmake_depends "Unix Makefiles" /home/wtwhite/win/repro_seqan_pthread_link_error /home/wtwhite/win/repro_seqan_pthread_link_error /home/wtwhite/win/repro_seqan_pthread_link_error/build /home/wtwhite/win/repro_seqan_pthread_link_error/build /home/wtwhite/win/repro_seqan_pthread_link_error/build/CMakeFiles/x.dir/DependInfo.cmake --color=
Dependee "/home/wtwhite/win/repro_seqan_pthread_link_error/build/CMakeFiles/x.dir/DependInfo.cmake" is newer than depender "/home/wtwhite/win/repro_seqan_pthread_link_error/build/CMakeFiles/x.dir/depend.internal".
Dependee "/home/wtwhite/win/repro_seqan_pthread_link_error/build/CMakeFiles/CMakeDirectoryInformation.cmake" is newer than depender "/home/wtwhite/win/repro_seqan_pthread_link_error/build/CMakeFiles/x.dir/depend.internal".
Scanning dependencies of target x
make[2]: warning:  Clock skew detected.  Your build may be incomplete.
make[2]: Leaving directory '/home/wtwhite/win/repro_seqan_pthread_link_error/build'
make -f CMakeFiles/x.dir/build.make CMakeFiles/x.dir/build
make[2]: Entering directory '/home/wtwhite/win/repro_seqan_pthread_link_error/build'
make[2]: Warning: File 'CMakeFiles/x.dir/flags.make' has modification time 0,28 s in the future
[ 50%] Building CXX object CMakeFiles/x.dir/x.cpp.o
/usr/bin/c++   -DSEQAN_HAS_EXECINFO=1 -DSEQAN_HAS_ZLIB=1 -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -I/home/wtwhite/win/seqan/include  -O3 -DNDEBUG   -o CMakeFiles/x.dir/x.cpp.o -c /home/wtwhite/win/repro_seqan_pthread_link_error/x.cpp
[100%] Linking CXX executable x
/usr/bin/cmake -E cmake_link_script CMakeFiles/x.dir/link.txt --verbose=1
/usr/bin/c++     -O3 -DNDEBUG   -Wl,--whole-archive -lpthread -Wl,--no-whole-archive CMakeFiles/x.dir/x.cpp.o  -o x -rdynamic -lrt -lz 
CMakeFiles/x.dir/x.cpp.o: In function `seqan::basic_unbgzf_streambuf<char, std::char_traits<char>, std::allocator<char>, char, std::allocator<char> >::basic_unbgzf_streambuf(std::istream&, unsigned long, unsigned long)':
x.cpp:(.text._ZN5seqan22basic_unbgzf_streambufIcSt11char_traitsIcESaIcEcS3_EC2ERSimm[_ZN5seqan22basic_unbgzf_streambufIcSt11char_traitsIcESaIcEcS3_EC5ERSimm]+0x66b): undefined reference to `pthread_create'
CMakeFiles/x.dir/x.cpp.o: In function `seqan::basic_bgzf_streambuf<char, std::char_traits<char>, std::allocator<char>, char, std::allocator<char> >::basic_bgzf_streambuf(std::ostream&, unsigned long, unsigned long)':
x.cpp:(.text._ZN5seqan20basic_bgzf_streambufIcSt11char_traitsIcESaIcEcS3_EC2ERSomm[_ZN5seqan20basic_bgzf_streambufIcSt11char_traitsIcESaIcEcS3_EC5ERSomm]+0x80e): undefined reference to `pthread_create'
collect2: error: ld returned 1 exit status
CMakeFiles/x.dir/build.make:95: recipe for target 'x' failed
make[2]: *** [x] Error 1
make[2]: Leaving directory '/home/wtwhite/win/repro_seqan_pthread_link_error/build'
CMakeFiles/Makefile2:67: recipe for target 'CMakeFiles/x.dir/all' failed
make[1]: *** [CMakeFiles/x.dir/all] Error 2
make[1]: Leaving directory '/home/wtwhite/win/repro_seqan_pthread_link_error/build'
Makefile:83: recipe for target 'all' failed
make: *** [all] Error 2
```

**MWE:** [repro_seqan_pthread_link_error.zip](https://github.com/seqan/seqan/files/1700518/repro_seqan_pthread_link_error.zip)

I hope the above helps, if not please let me know!

--- a/apps/mason2/CMakeLists.txt
+++ b/apps/mason2/CMakeLists.txt
@@ -103,14 +103,14 @@ add_executable (mason_splicing mason_splicing.cpp)
 add_executable (mason_tests mason_tests.cpp)
 
 # Add dependencies found by find_package (SeqAn).
-target_link_libraries (mason_frag_sequencing ${SEQAN_LIBRARIES} mason_sim)
-target_link_libraries (mason_genome ${SEQAN_LIBRARIES} mason_sim)
-target_link_libraries (mason_materializer ${SEQAN_LIBRARIES} mason_sim)
-target_link_libraries (mason_methylation ${SEQAN_LIBRARIES} mason_sim)
-target_link_libraries (mason_simulator ${SEQAN_LIBRARIES} mason_sim)
-target_link_libraries (mason_splicing ${SEQAN_LIBRARIES} mason_sim)
-target_link_libraries (mason_tests ${SEQAN_LIBRARIES} mason_sim)
-target_link_libraries (mason_variator ${SEQAN_LIBRARIES} mason_sim)
+target_link_libraries (mason_frag_sequencing mason_sim)
+target_link_libraries (mason_genome mason_sim)
+target_link_libraries (mason_materializer mason_sim)
+target_link_libraries (mason_methylation mason_sim)
+target_link_libraries (mason_simulator mason_sim)
+target_link_libraries (mason_splicing mason_sim)
+target_link_libraries (mason_tests mason_sim)
+target_link_libraries (mason_variator mason_sim)
 
 # ----------------------------------------------------------------------------
 # Setup Common Tool Description for Generic Workflow Nodes

--- a/util/cmake/seqan-config.cmake
+++ b/util/cmake/seqan-config.cmake
@@ -307,7 +307,7 @@ endif ()
 
 # some OSes don't link pthread fully when building statically so we explicitly include whole archive
 if (UNIX AND NOT APPLE)
-    set (CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -Wl,--whole-archive -lpthread -Wl,--no-whole-archive")
+    set (SEQAN_LIBRARIES ${SEQAN_LIBRARIES} "-Wl,--whole-archive -lpthread -Wl,--no-whole-archive")
 endif ()
 
 if ((${CMAKE_SYSTEM_NAME} STREQUAL "FreeBSD") OR (${CMAKE_SYSTEM_NAME} STREQUAL "OpenBSD"))

